### ISSUE TITLE
node crash when running wechaty in wine

### ISSUE THREAD

**Issue by hcfw007** [huan/docker-wine#6]
Hola.
I'm not sure this repo is the correct to post this issue. I think it's more of a wine issue than a wechaty issue.

As we talked before, I'm trying to upgrade wechaty-puppet-wxwork to node 16. So far, I can start wechaty-io-client, and wxwork.exe with wine 6.20. It's the staging branch, however wine 6.0.2 stable does not have the WS2_32 support I need.
After about 10 mins (image created at 14:39:30, last wechty log at 14:49:29 ), node will crash. Last function called in wechaty is ws.ping().
```
14:48:59 SILL Io start() setInterval() ws.ping()
14:49:01 INFO Bridge onEvent() type: 11087(MT_HEARTBEAT), data: {"data":{},"type
":11087}
14:49:01 SILL Manager throttleQueue emit heartbeat.
14:49:01 SILL Io send({"name":"heartbeat","payload":{"cuid":"ckvj5i1020000csn2ag
kq8q84","data":"[object Object]"}})
14:49:01 SILL Wechaty memoryCheck() free: 17946 MB, require: 4 MB
14:49:01 SILL Manager heartbeat event received
14:49:02 SILL Io initWebSocket() ws.on(message): {"name":"sys","payload":"roger"
}
0274:fixme:winsock:WSAIoctl unimplemented ioctl SIO_IDEAL_SEND_BACKLOG_QUERY
0274:fixme:winsock:server_ioctl_sock Unsupported ioctl 120348 (device=12 access=0 func=d2 method=0)
14:49:09 SILL Io start() setInterval() ws.ping()
14:49:11 INFO Bridge onEvent() type: 11087(MT_HEARTBEAT), data: {"data":{},"type
":11087}
14:49:11 SILL Manager heartbeat event received
0274:fixme:winsock:WSAIoctl unimplemented ioctl SIO_IDEAL_SEND_BACKLOG_QUERY
0274:fixme:winsock:server_ioctl_sock Unsupported ioctl 120348 (device=12 access=0 func=d2 method=0)
14:49:19 SILL Io start() setInterval() ws.ping()
14:49:22 INFO Bridge onEvent() type: 11087(MT_HEARTBEAT), data: {"data":{},"type
":11087}
14:49:22 SILL Manager throttleQueue emit heartbeat.
14:49:22 SILL Io send({"name":"heartbeat","payload":{"cuid":"ckvj5i1020000csn2ag
kq8q84","data":"[object Object]"}})
14:49:22 SILL Wechaty memoryCheck() free: 17944 MB, require: 4 MB
14:49:22 SILL Manager heartbeat event received
14:49:22 SILL Io initWebSocket() ws.on(message): {"name":"sys","payload":"roger"
}
0274:fixme:winsock:WSAIoctl unimplemented ioctl SIO_IDEAL_SEND_BACKLOG_QUERY
0274:fixme:winsock:server_ioctl_sock Unsupported ioctl 120348 (device=12 access=0 func=d2 method=0)
14:49:29 SILL Io start() setInterval() ws.ping()
01d0:err:sync:RtlpWaitForCriticalSection section 05266350 "?" wait timed out in thread 01d0, blocked by 0000, retrying (60 sec)
0274:fixme:winsock:WSAIoctl unimplemented ioctl SIO_IDEAL_SEND_BACKLOG_QUERY
0274:fixme:winsock:server_ioctl_sock Unsupported ioctl 120348 (device=12 access=0 func=d2 method=0)
0218:err:sync:RtlpWaitForCriticalSection section 052FFD14 "?" wait timed out in thread 0218, blocked by 01d0, retrying (60 sec)
0274:fixme:winsock:WSAIoctl unimplemented ioctl SIO_IDEAL_SEND_BACKLOG_QUERY
0274:fixme:winsock:server_ioctl_sock Unsupported ioctl 120348 (device=12 access=0 func=d2 method=0)
```

However ws.ping() is not the problem. I quoted the line and leaves the log, node still crashes after `SILL Io start() setInterval() ws.ping()`

Now I'm out of clues. Have you got any idea what could be wrong?

**Comment by huan** [comment]
Hi it's very hard for me to just say something by reading your short log messages, you need to provide more details with useful information so that I can know what happens.

And I'm curious about what is `WS2_32`, and why it becomes your blocker? I think the wine 5.x has already been good enough for running any win32 programs, we should keep things as simple as we can.

**Comment by hcfw007** [comment]
Because in wine 5.0, running wechaty-io-client will encounter WS2_32.dll.GetHostNameW unimplemented error.

```
> juzi-work@1.0.0 start
> wechaty-io-client

0059:fixme:winsock:WS_EnterSingleProtocolW unknown Protocol <0x00000000>
0059:fixme:winsock:WS_EnterSingleProtocolW unknown Protocol <0x00000000>
005a:fixme:process:NtQueryInformationProcess (0xffffffff,info_class=49,0x3e9fe5c,0x00000004,(nil)) Unknown information class
005c:fixme:kernelbase:AppPolicyGetThreadInitializationType FFFFFFFA, 040BFF04
0059:fixme:ntdll:NtQuerySystemInformation info_class SYSTEM_PERFORMANCE_INFORMATION
0059:fixme:ntdll:EtwEventRegister ({77754e9b-264b-4d8d-b981-e4135c1ecb0c}, 0x4b3fb0, (nil), 0x3737970) stub.
0059:fixme:ntdll:NtQueryVolumeInformationFile 0xbc: volume info not supported
125.33.25.82 - - [04/Nov/2021 10:25:11] 125.33.25.82: Plain non-SSL (ws://) WebSocket connection
125.33.25.82 - - [04/Nov/2021 10:25:11] 125.33.25.82: Version hybi-13, base64: 'False'
125.33.25.82 - - [04/Nov/2021 10:25:11] 125.33.25.82: Path: '/websockify'
125.33.25.82 - - [04/Nov/2021 10:25:11] connecting to: 127.0.0.1:5900
10:25:12 SILL Puppet Config: WECHATY_LOG set level to silly
wine: Call from 0x7bc78ce1 to unimplemented function WS2_32.dll.GetHostNameW, aborting
wine: Unimplemented function WS2_32.dll.GetHostNameW called at address 7BC78CE1 (thread 0059), starting debugger...
004a:fixme:winsock:WS_setsockopt Unknown IPPROTO_TCP optname 0x00000003
```

detailed log in GUI:
```
Unhandled exception: unimplemented function WS2_32.dll.GetHostNameW called in 32-bit code (0x7bc78ce1).
Register dump:
 CS:0023 SS:002b DS:002b ES:002b FS:0063 GS:006b
 EIP:7bc78ce1 ESP:0032e550 EBP:0032e5c4 EFLAGS:00200206(   - --  I   - -P- )
 EAX:037181e6 EBX:0032e568 ECX:00000090 EDX:03c63fc0
 ESI:03c6f3c0 EDI:0032e808
Stack dump:
0x0032e550:  00000000 03c63fc0 0032e5d4 03c690f0
0x0032e560:  03c63fc0 00be3e0a 80000100 00000001
0x0032e570:  00000000 7bc78ce1 00000002 037182bc
0x0032e580:  037181e6 00be463d 0032e5d4 00000044
0x0032e590:  00000001 00bf2d11 060bc881 00000000
0x0032e5a0:  00000003 060bc862 0589da18 0000001f
Backtrace:
=>0 0x7bc78ce1 call_dll_entry_point+0x591() in ntdll (0x0032e5c4)
  1 0x0033000f (0x0032e810)
  2 0x0032e824 (0x00a9befb)
  3 0x5e74c085 (0x0424448b)
0x7bc78ce1 call_dll_entry_point+0x591 in ntdll: addl	$12,%esp
Modules:
Module	Address			Debug info	Name (65 modules)
PE	  400000- 3a6e000	Deferred        node
PE	61540000-6154b000	Deferred        api-ms-win-core-localization-obsolete-l1-2-0
PE	639c0000-63b30000	Deferred        winmm
PE	65880000-6588c000	Deferred        api-ms-win-core-sysinfo-l1-2-1
PE	65980000-6599d000	Deferred        version
PE	66640000-666a3000	Deferred        msacm32
PE	66800000-6680b000	Deferred        api-ms-win-core-datetime-l1-1-1
PE	66980000-6698b000	Deferred        api-ms-win-core-string-l1-1-0
PE	67f00000-67f24000	Deferred        userenv
PE	68ac0000-68ae0000	Deferred        powrprof
PE	6aec0000-6aece000	Deferred        psapi
PE	6ba80000-6ba8b000	Deferred        api-ms-win-core-fibers-l1-1-1
PE	6c0c0000-6c107000	Deferred        imm32
PE	6c300000-6c30c000	Deferred        api-ms-win-appmodel-runtime-l1-1-2
PE	6e540000-6e54c000	Deferred        api-ms-win-core-synch-l1-2-0
PE	6e8c0000-6e8cc000	Deferred        api-ms-win-core-localization-l1-2-1
PE	6fdc0000-6ffee000	Deferred        rpcrt4
PE	774e0000-7761d000	Deferred        ole32
PE	7b000000-7b2e9000	Deferred        kernelbase
ELF	7b400000-7b673000	Deferred        kernel32<elf>
  \-PE	7b420000-7b673000	\               kernel32
ELF	7bc00000-7beb8000	Dwarf           ntdll<elf>
  \-PE	7bc30000-7beb8000	\               ntdll
ELF	7c000000-7c005000	Deferred        <wine-loader>
ELF	7dac3000-7db50000	Deferred        libgmp.so.10
ELF	7db50000-7dcd2000	Deferred        libunistring.so.2
ELF	7dcd2000-7de29000	Deferred        libp11-kit.so.0
ELF	7dea4000-7dec3000	Deferred        libgcc_s.so.1
ELF	7dec3000-7decd000	Deferred        libffi.so.6
ELF	7decd000-7df06000	Deferred        libhogweed.so.4
ELF	7df06000-7df43000	Deferred        libnettle.so.6
ELF	7df43000-7df5a000	Deferred        libtasn1.so.6
ELF	7df5a000-7df7a000	Deferred        libidn2.so.0
ELF	7df7a000-7e15f000	Deferred        libgnutls.so.30
ELF	7e189000-7e193000	Deferred        libuuid.so.1
ELF	7e193000-7e1bf000	Deferred        libexpat.so.1
ELF	7e1bf000-7e20c000	Deferred        libfontconfig.so.1
ELF	7e20c000-7e24b000	Deferred        libpng16.so.16
ELF	7e24b000-7e30d000	Deferred        libfreetype.so.6
ELF	7e30d000-7e332000	Deferred        bcrypt<elf>
  \-PE	7e320000-7e332000	\               bcrypt
ELF	7e332000-7e42b000	Deferred        crypt32<elf>
  \-PE	7e350000-7e42b000	\               crypt32
ELF	7e42b000-7e585000	Deferred        gdi32<elf>
  \-PE	7e450000-7e585000	\               gdi32
ELF	7e585000-7e7c7000	Deferred        user32<elf>
  \-PE	7e5c0000-7e7c7000	\               user32
ELF	7e7c7000-7e8a1000	Deferred        msvcrt<elf>
  \-PE	7e7f0000-7e8a1000	\               msvcrt
ELF	7e8a1000-7e926000	Deferred        advapi32<elf>
  \-PE	7e8c0000-7e926000	\               advapi32
ELF	7e926000-7e954000	Deferred        iphlpapi<elf>
  \-PE	7e930000-7e954000	\               iphlpapi
ELF	7e954000-7e992000	Deferred        ws2_32<elf>
  \-PE	7e960000-7e992000	\               ws2_32
ELF	7e992000-7e9b0000	Deferred        libz.so.1
ELF	7e9b0000-7ea38000	Deferred        dbghelp<elf>
  \-PE	7e9d0000-7ea38000	\               dbghelp
ELF	7ea38000-7ea4d000	Deferred        libnss_files.so.2
ELF	7eefb000-7f000000	Deferred        libm.so.6
ELF	f7b2a000-f7b30000	Deferred        libdl.so.2
ELF	f7b30000-f7d1d000	Deferred        libc.so.6
ELF	f7d1d000-f7d40000	Deferred        libpthread.so.0
ELF	f7d4d000-f7f01000	Dwarf           libwine.so.1
ELF	f7f03000-f7f2e000	Deferred        ld-linux.so.2
Threads:
process  tid      prio (all id:s are in hex)
00000008 CMD.EXE
	00000009    0
0000000e services.exe
	00000024    0
	0000001a    0
	00000013    0
	00000010    0
	0000000f    0
00000011 plugplay.exe
	00000017    0
	00000016    0
	00000012    0
00000018 winedevice.exe
	00000021    0
	0000001d    0
	0000001c    0
	0000001b    0
	00000019    0
0000001e explorer.exe
	00000020    0
	0000001f    0
00000022 winedevice.exe
	00000026    0
	00000025    0
	00000023    0
0000002a AutoIt3.exe
	0000002b    0
0000002d CMD.EXE
	0000002e    0
00000049 node.exe
	00000055    0
	00000054    0
	00000053    0
	00000052    0
	00000051    0
	00000050    0
	0000004f    0
	0000004e    0
	0000004d    0
	0000004b    0
	0000004a    0
00000056 cmd.exe
	00000057    0
00000058 (D) C:\node\node.exe
	00000060    0
	0000005f    0
	0000005e    0
	0000005d    0
	0000005c    0
	0000005b    0
	0000005a    0
	00000059    0 <==
System information:
    Wine build: wine-5.0
    Platform: i386
    Version: Windows 7
    Host system: Linux
```

**Comment by huan** [comment]
I guess that that warning message will not block the program from running at all?

> unimplemented function WS2_32.dll.GetHostNameW

Please correct me if I'm not right, but I don't think it's a big deal if it's just a warning message.

**Comment by hcfw007** [comment]
Yes it will, VNC will show a wine dialog saying node crashed.
![QQ图片20211105010708](https://user-images.githubusercontent.com/13669999/140386017-69d1fb42-ce22-43eb-b781-0b2c99a3adb3.png)
![QQ图片20211105010715](https://user-images.githubusercontent.com/13669999/140386026-ffcddce9-87c8-4070-bd5c-e90597d028eb.png)
 
The detailed info in the second image is in the comment above.

**Comment by hcfw007** [comment]
As far as I know, any log saying 'starting debugger...' will block the program.

**Comment by huan** [comment]
Oh, that's really bad news.

I hope you can solve it soon.

**Comment by hcfw007** [comment]
Sad face.
Github does not have a sad face emoji. Now I'm even sadder.

**Comment by hcfw007** [comment]
I reproduced this crash in ubuntu VM, and make a little progress. In VM the crash log can be displayed.
```
Unhandled exception: page fault on read access to 0x00000001 in 32-bit code (0x00d5d13d).
Register dump:
 CS:0023 SS:002b DS:002b ES:002b FS:0063 GS:006b
 EIP:00d5d13d ESP:03b6f450 EBP:03b6f490 EFLAGS:00210246(  R- --  I  Z- -P- )
 EAX:00000001 EBX:03bdf928 ECX:00000001 EDX:00000000
 ESI:00000011 EDI:00000000
Stack dump:
0x03b6f450:  00000002 03b6f4d4 03b6f4c8 03bdf928
0x03b6f460:  00000000 0e2020d5 00000044 0f849701
0x03b6f470:  00000018 04d03c7d 00000011 00000438
0x03b6f480:  00000000 04d098d1 0e2020d5 0000001c
0x03b6f490:  03b6f4b8 00df8824 04e4232d 04e4232d
0x03b6f4a0:  00000000 0000003a 03b6f4ec 03c1c2d8
Backtrace:
=>0 0x00d5d13d ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x2fcd0d() in node (0x03b6f490)
  1 0x00df8824 ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x3983f3() in node (0x03b6f4b8)
  2 0x00d2a216 ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x2c9de5() in node (0x03b6f4ec)
  3 0x00d2a216 ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x2c9de5() in node (0x03b6f520)
  4 0x00d2a216 ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x2c9de5() in node (0x03b6f564)
  5 0x00d2a216 ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x2c9de5() in node (0x03b6f598)
  6 0x00d2a216 ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x2c9de5() in node (0x03b6f5c4)
  7 0x00d2a216 ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x2c9de5() in node (0x03b6f618)
  8 0x00d2895c ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x2c852b() in node (0x03b6f634)
  9 0x00d28789 ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x2c8358() in node (0x03b6f660)
  10 0x00c2b010 ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x1cabdf() in node (0x03b6f6dc)
  11 0x03c34274 (0x03c3427c)
  12 0x04d020a1 (0x128db461)
  13 0xa504e421 (0xa504cc21)
0x00d5d13d ??$Add@VIsolate@internal@v8@@@?$Dictionary@VGlobalDictionary@internal@v8@@VGlobalDictionaryShape@23@@internal@v8@@SA?AV?$Handle@VGlobalDictionary@internal@v8@@@12@PAVIsolate@12@V312@V?$Handle@VName@internal@v8@@@12@V?$Handle@VObject@internal@v8@@@12@VProp+0x2fcd0d in node: movzbl	0x0(%eax,%edx,1),%eax
Modules:
Module	Address			Debug info	Name (77 modules)
PE	  110000-  130000	Deferred        userenv
PE	  400000- 3a6e000	Export          node
PE	 3ca0000- 3de4000	Deferred        winmm
PE	 e510000- e53f000	Deferred        ffi_bindings.node
PE	103c0000-1043a000	Deferred        node.napi.node
PE	10970000-10a9c000	Deferred        node_sqlite3.node
PE	10c80000-115a6000	Deferred        wxworkloaderjs_x86
PE	61500000-6156d000	Deferred        iphlpapi
PE	61740000-61801000	Deferred        advapi32
PE	62b80000-62b8d000	Deferred        api-ms-win-core-localization-l1-2-1
PE	62fc0000-63193000	Deferred        rpcrt4
PE	63480000-63498000	Deferred        version
PE	63bc0000-63bfa000	Deferred        shcore
PE	64a40000-64b3e000	Deferred        shlwapi
PE	65580000-655a2000	Deferred        wtsapi32
PE	66240000-66276000	Deferred        bcrypt
PE	66b00000-66b50000	Deferred        win32u
PE	66e40000-66e4d000	Deferred        api-ms-win-appmodel-runtime-l1-1-2
PE	67800000-6780c000	Deferred        api-ms-win-core-string-l1-1-0
PE	67840000-6784f000	Deferred        psapi
PE	69880000-6988c000	Deferred        api-ms-win-core-fibers-l1-1-0
PE	6a040000-6a090000	Deferred        msacm32
PE	6a280000-6a46a000	Deferred        msvcrt
PE	6ac80000-6ac8c000	Deferred        api-ms-win-core-fibers-l1-1-1
PE	6b500000-6b50c000	Deferred        api-ms-win-core-localization-obsolete-l1-2-0
PE	6bc00000-6bc80000	Deferred        sechost
PE	6bcc0000-6be41000	Deferred        setupapi
PE	6c9c0000-6cb48000	Deferred        gdi32
PE	6cec0000-6cedb000	Deferred        powrprof
PE	6d780000-6d7e3000	Deferred        ws2_32
PE	6de80000-6de8d000	Deferred        api-ms-win-core-synch-l1-2-0
PE	6e4c0000-6e4f3000	Deferred        dnsapi
PE	6e8c0000-6eadf000	Deferred        crypt32
PE	6ed00000-6f3b3000	Deferred        user32
PE	6f480000-6f49b000	Deferred        nsi
PE	6fa80000-6fc79000	Deferred        dbghelp
PE	6ff80000-6ff8c000	Deferred        api-ms-win-core-datetime-l1-1-1
PE	700c0000-700cd000	Deferred        api-ms-win-core-sysinfo-l1-2-1
PE	70b40000-70d7c000	Deferred        ucrtbase
PE	71200000-7123e000	Deferred        imm32
PE	774e0000-7761d000	Deferred        ole32
PE	7b000000-7b274000	Deferred        kernelbase
PE	7b600000-7b8ed000	Deferred        kernel32
PE	7bc00000-7be14000	Deferred        ntdll
ELF	7d000000-7d005000	Deferred        <wine-loader>
ELF	7e05d000-7e07c000	Deferred        libgcc_s.so.1
ELF	7e07c000-7e084000	Deferred        libnss_dns.so.2
ELF	7e084000-7e099000	Deferred        libnss_files.so.2
ELF	7e0af000-7e0b9000	Deferred        libffi.so.7
ELF	7e0b9000-7e149000	Deferred        libgmp.so.10
ELF	7e149000-7e182000	Deferred        libhogweed.so.5
ELF	7e182000-7e1c2000	Deferred        libnettle.so.7
ELF	7e1c2000-7e1d9000	Deferred        libtasn1.so.6
ELF	7e1d9000-7e35b000	Deferred        libunistring.so.2
ELF	7e35b000-7e37b000	Deferred        libidn2.so.0
ELF	7e37b000-7e4d3000	Deferred        libp11-kit.so.0
ELF	7e4d3000-7e6d3000	Deferred        libgnutls.so.30
ELF	7e74e000-7e758000	Deferred        libuuid.so.1
ELF	7e758000-7e784000	Deferred        libexpat.so.1
ELF	7e784000-7e7d2000	Deferred        libfontconfig.so.1
ELF	7e7d2000-7e7f0000	Deferred        libz.so.1
ELF	7e7f0000-7e82f000	Deferred        libpng16.so.16
ELF	7e82f000-7e8f5000	Deferred        libfreetype.so.6
ELF	7e8f5000-7e90d000	Deferred        crypt32.so
ELF	7e90d000-7e9ba000	Deferred        win32u.so
ELF	7e9ba000-7e9d3000	Deferred        libresolv.so.2
ELF	7e9e9000-7ea01000	Deferred        dnsapi.so
ELF	7ea01000-7eb87000	Dwarf           libwine.so.1
ELF	f7b67000-f7c6c000	Deferred        libm.so.6
ELF	f7c6c000-f7c77000	Deferred        librt.so.1
ELF	f7c77000-f7d39000	Deferred        ntdll.so
ELF	f7d39000-f7f28000	Deferred        libc.so.6
ELF	f7f28000-f7f2e000	Deferred        libdl.so.2
ELF	f7f2e000-f7f51000	Deferred        libpthread.so.0
ELF	f7f55000-f7f5f000	Deferred        bcrypt.so
ELF	f7f5f000-f7f65000	Deferred        ws2_32.so
ELF	f7f69000-f7f96000	Deferred        ld-linux.so.2
Threads:
process  tid      prio (all id:s are in hex)
00000038 services.exe
	["C:\windows\system32\services.exe"]
	0000003c    0
	00000040    0
	0000004c    0
	00000070    0
	0000009c    0
	000000b4    0
00000044 winedevice.exe
	[C:\windows\system32\winedevice.exe]
	00000048    0
	00000054    0
	00000058    0
	0000005c    0
	000000c0    0
00000060 winedevice.exe
	[C:\windows\system32\winedevice.exe]
	00000064    0
	00000074    0
	00000078    0
	0000007c    0
	00000084    0
	00000088    0
	0000008c    0
00000094 plugplay.exe
	[C:\windows\system32\plugplay.exe]
	00000098    0
	000000a0    0
	000000a4    0
	000000a8    0
000000ac svchost.exe
	[C:\windows\system32\svchost.exe -k LocalServiceNetworkRestricted]
	000000b0    0
	000000b8    0
	000000bc    0
00000178 start.exe
	["C:\windows\system32\start.exe" /exec CMD.EXE /C /tmp/winescript_42713_1636442659149305755.bat]
	0000017c    0
00000180 conhost.exe
	["C:\windows\system32\conhost.exe" --unix --width 77 --height 21 --server 0xc]
	00000184    0
00000188 explorer.exe
	[C:\windows\system32\explorer.exe /desktop]
	0000018c    0
	00000190    0
00000194 cmd.exe
	["C:\windows\system32\CMD.EXE" /C /tmp/winescript_42713_1636442659149305755.bat]
	00000198    0
000001fc node.exe
	["C:\node\\node.exe" "C:\node\node_modules\npm\bin\npm-cli.js" start]
	00000200    0
	00000204    0
	00000208    0
	0000020c    0
	00000210    0
	00000214    0
	00000218    0
	0000021c    0
	00000220    0
	00000224    0
	00000228    0
	00000234    0
0000022c cmd.exe
	[C:\windows\system32\cmd.exe /d /s /c wechaty-io-client]
	00000230    0
00000238 (D) C:\node\node.exe
	["node"  "C:\juzi\node_modules\.bin\\..\wechaty\dist\bin\io-client.js" ]
	0000023c    0 <==
	00000244    0
	00000248    0
	0000024c    0
	00000250    0
	00000254    0
	00000258    0
	0000025c    0
	00000260    0
	00000264    0
	00000268    0
	0000026c    0
	00000270    0
	00000274    0
	00000278    0
	0000027c    0
	00000280    0
	00000284    0
	00000288    0
	0000028c    0
	0000031c    0
00000290 WXWork.exe
	["C:\Program Files (x86)\WXWork\WXWork.exe"]
	00000294    0
	00000298    0
	0000029c    0
	000002a0    0
	000002a4    0
	000002a8    0
	000002ac    0
	000002b0    0
	000002b4    0
	000002b8    0
	000002bc    0
	000002c0    0
	000002c4    0
	000002c8    0
	000002cc    0
	000002d0    0
	000002d4    0
	000002d8    0
	000002dc    0
	000002e0    0
	000002e4    0
	000002e8    0
	000002ec    0
	000002f0    0
	000002f4    0
	000002f8    0
	000002fc    0
	00000300    0
	00000304    0
	00000308    0
	00000310    0
00000320 explorer.exe
	[C:\windows\system32\explorer.exe /desktop]
	00000324    0
	00000328    0
System information:
    Wine build: wine-6.20 (Staging)
    Platform: i386
    Version: Windows 10
    Host system: Linux
    Host version: 5.11.0-38-generic

```

**Comment by huan** [comment]
Question: have you tried to run `wechaty-getting-started` repo with the web protocol?

1. If so, can it work as expected? 
2. If not, please try it and to see if you can find some new information.

**Comment by hcfw007** [comment]
Thank you for your advice. I have started a ding-dong bot running in ubuntu VM and so far it's working well. (over 30 mins)

**Comment by huan** [comment]
It's great to know that you have confirmed the Wine 6, Node.js 16, and Wechaty all work without any problem!

I hope this fact can help you reduce the search domain and be more focused on the real problem.

**Comment by hcfw007** [comment]
Do you have any idea about debugging wine?
So far wxwork puppet works well on Win10 VM, so something is not working with wine.
I'm trying to find this 'something' but have no idea about how to find it.

**Comment by huan** [comment]
In order to catch the `thing`, you need to split your source code and finger out which part cause the system crash.

That is called `minimum reproducible code`.

**Comment by hcfw007** [comment]
Updates: I have narrow down the suspects to the dll or node-ffi module. Does the log above tell us anything about it? I don't really understand wine logs.

**Comment by huan** [comment]
Unfortunately, I know nothing more than you about the wine logs.

However, I think you can start to make the **REAL** minimum reproducible code for helping you understand this problem better.

Here's how to do it:

1. Create a js file with the fewest lines of code as possible as you can
2. without loading any unnecessary modules
3. with an easy to reproducible environment
4. reproduce the crash in no more than 10 seconds

And keep this in mind: your problem is neither wine 6 crash, nor node.js v16 crash, you have to find the very specific code that makes you crash, and only that point is meaningful for debugging.

**Comment by hcfw007** [comment]
Thank you for your suggestion. Actually that's exactly what I have done. I have written a test script to do the following:
1. bind wxworkloader and wxworkhelper dll
2. open wxwork.exe
3. display any event with console

And that leads to corruption of node. Unfortunately so far I can not separate this code further to see exactly which part of the problem is. My guess is  some memory address when dll emitting events.

**Comment by hcfw007** [comment]
Updates: Problem somehow disappeared after upgrading wine to 6.21.
